cpp/include/raft/sparse/distance/l2_distance.cuh

```diff
@@ -108,6 +148,59 @@ void compute_l2(value_t *out, const value_idx *Q_coo_rows,
                     expansion_func);
 }
 
+template <typename value_idx, typename value_t, int tpb = 256>
+void compute_corr(value_t *out, const value_idx *Q_coo_rows,
+                  const value_t *Q_data, value_idx Q_nnz,
+                  const value_idx *R_coo_rows, const value_t *R_data,
+                  value_idx R_nnz, value_idx m, value_idx n, value_idx n_cols,
+                  std::shared_ptr<raft::mr::device::allocator> alloc,
+                  cudaStream_t stream) {
+  // sum_sq for std dev
+  raft::mr::device::buffer<value_t> Q_sq_norms(alloc, stream, m);
+  raft::mr::device::buffer<value_t> R_sq_norms(alloc, stream, n);
+
+  // sum for mean
+  raft::mr::device::buffer<value_t> Q_norms(alloc, stream, m);
+  raft::mr::device::buffer<value_t> R_norms(alloc, stream, n);
+
+  CUDA_CHECK(
+    cudaMemsetAsync(Q_sq_norms.data(), 0, Q_sq_norms.size() * sizeof(value_t)));
+  CUDA_CHECK(
+    cudaMemsetAsync(R_sq_norms.data(), 0, R_sq_norms.size() * sizeof(value_t)));
+
+  CUDA_CHECK(
+    cudaMemsetAsync(Q_norms.data(), 0, Q_norms.size() * sizeof(value_t)));
+  CUDA_CHECK(
+    cudaMemsetAsync(R_norms.data(), 0, R_norms.size() * sizeof(value_t)));
+
+  compute_row_norm_kernel<<<raft::ceildiv(Q_nnz, tpb), tpb, 0, stream>>>(
+    Q_sq_norms.data(), Q_coo_rows, Q_data, Q_nnz);
+  compute_row_norm_kernel<<<raft::ceildiv(R_nnz, tpb), tpb, 0, stream>>>(
+    R_sq_norms.data(), R_coo_rows, R_data, R_nnz);
+
+  compute_row_sum_kernel<<<raft::ceildiv(Q_nnz, tpb), tpb, 0, stream>>>(
+    Q_norms.data(), Q_coo_rows, Q_data, Q_nnz);
+  compute_row_sum_kernel<<<raft::ceildiv(R_nnz, tpb), tpb, 0, stream>>>(
+    R_norms.data(), R_coo_rows, R_data, R_nnz);
+
+  value_t n_cols_div = 1.0 / n_cols;
+  raft::linalg::unaryOp<value_t>(
+    Q_sq_norms.data(), Q_sq_norms.data(), m,
+    [=] __device__(value_t input) { return input * n_cols_div; }, stream);
```

Comment:
Extract repeating lambdas into one lambda. Also, these kernels could be combined, right? Using `thrust::zip_iterator` and `thrust::transform`, only one kernel launch will be needed to do these 4 things

Reply:
I'm not sure I understand what you are saying here. I'm using our unary op kernel to perform an element-wise operation. I  generally prefer this over thrust::transforms just because I like the layer of indirection in a consistent API (for all I care, we could just call thrust underneath). However, I'm not sure I agree with your statement about a single kernel launch to compute row-wise norms and combine them with a dot product (e.g. I'm not sure I'd want a single kernel to do all of that). I might be misunderstanding what you mean, though.

Reply:
Ah, my mistake. The changelog w/ your comment wasn't showing me all the unary op calls below. I don't mind combining those.